clippy_lints/src/checked_conversions.rs

```diff
@@ -50,61 +50,62 @@ impl_lint_pass!(CheckedConversions => [CHECKED_CONVERSIONS]);
 
 impl<'tcx> LateLintPass<'tcx> for CheckedConversions {
     fn check_expr(&mut self, cx: &LateContext<'_>, item: &Expr<'_>) {
-        if !self.msrv.meets(msrvs::TRY_FROM) {
-            return;
-        }
-
-        let result = if !in_constant(cx, item.hir_id)
-            && !in_external_macro(cx.sess(), item.span)
-            && let ExprKind::Binary(op, left, right) = &item.kind
-        {
-            match op.node {
-                BinOpKind::Ge | BinOpKind::Le => single_check(item),
-                BinOpKind::And => double_check(cx, left, right),
-                _ => None,
+        if let ExprKind::Binary(op, lhs, rhs) = item.kind
+            && let (lt1, gt1, op2) = match op.node {
+                BinOpKind::Le => (lhs, rhs, None),
+                BinOpKind::Ge => (rhs, lhs, None),
+                BinOpKind::And
+                    if let ExprKind::Binary(op1, lhs1, rhs1) = lhs.kind
+                        && let ExprKind::Binary(op2, lhs2, rhs2) = rhs.kind =>
+                {
+                    let (lt1, gt1) = match op1.node {
+                        BinOpKind::Le => (lhs1, rhs1),
+                        BinOpKind::Ge => (rhs1, lhs1),
+                        _ => return,
+                    };
+                    let (lt2, gt2) = match op2.node {
+                        BinOpKind::Le => (lhs2, rhs2),
+                        BinOpKind::Ge => (rhs2, lhs2),
+                        _ => return,
+                    };
```

Comment:
without the `normalize_le_ge` function at:

https://github.com/rust-lang/rust-clippy/blob/e7f2952f7e0ea9758e0830fb854dd6faea38c1c2/clippy_lints/src/checked_conversions.rs#L312-L319

 this part becomes a bit harder to understand, and it seems repetitive, could we bring that function back? Or add some comments here, which also works fine by me.

Reply:
I'm going to push against this and then I'm going to turn around and agree with the conclusion anyways. This will be more of an argument against the idea that factoring something out into a function will help with readability.

So with this moved into a function we first read:
```rust
&& let Some((lt1, gt1)) = normalize_le_ge(op1.node, lhs1, rhs1)
```
Reading through this we can gleam two things quickly: we only continue if we get a pair of values, and the operation in question works on all the parts of a binary expression. Going by the name of the function mentioning `le` and `gt`, the results being named `lt1` and `gt1`; and the fact this works on a binary expression we can infer that we're doing something with the `<=` and `>=` operators and the result has to do with expressions on the less than and greater than sides (and not the left and right sides). With `normalize` in the name we can also infer that this probably normalizes something about the operator itself, which in this case is probably their order. This just leaves why the result is an `Option`. We can reasonably assume that `None` represents normalization failed which then leads to the question of why it would fail. We can continue from the previous inference of it's name that it works on the `<=` and `>=` operators, but not any others, which means there's a good chance failure refers to such a case. With that we now have an idea based on a pile of inference as to what that line of code actually does.

You'll notice throughout the previous walkthrough of reading the line not once was it mentioned exactly what the function does, only what we could gleam from the call site. Since this is our first encounter with that function we don't immediately know what it does. Though we could infer everything correctly in this case, it's still something which takes time. This little bit of extra time on inference is what will bring it in line with the extra time it takes to read combined with the fact that we still haven't confirmed what the function does leaves it without much advantage compared to the inlined form. Note we still don't know what the function does so this leaves us at an overall disadvantage.

When it comes to reading the second instance things change. Neither will be hard to spot the duplication in form, but spotting this across four lines vs. one line has a clear winner. The function call form also has the advantage that you could read both lines almost at the same time, which you're less likely to be able to do with the inlined form. This is probably enough to overcome the disadvantages, but not really much more than that.

With the argument against doing that out of the way, there is another benefit to be gained. This requires looking at the surrounding match expression:
```rust
match op.node {
    BinOpKind::Le => (lhs, rhs, None),
    BinOpKind::Ge => (rhs, lhs, None),
    BinOpKind::And
        if let ExprKind::Binary(op1, lhs1, rhs1) = lhs.kind
            && let ExprKind::Binary(op2, lhs2, rhs2) = rhs.kind =>
    {
        let (lt1, gt1) = match op1.node {
            BinOpKind::Le => (lhs1, rhs1),
            BinOpKind::Ge => (rhs1, lhs1),
            _ => return,
        };
        let (lt2, gt2) = match op2.node {
            BinOpKind::Le => (lhs2, rhs2),
            BinOpKind::Ge => (rhs2, lhs2),
            _ => return,
        };
        (lt1, gt1, Some((lt2, gt2)))
    },
    _ => return,
}
```
With the inlined form, this becomes harder to quickly scan due to the size of the `And` arm. This makes it harder to see how the results vary between the arms (`(_, _, None)` vs `(_, _, Some(_))`) and harder to see the total list of operators matched. For completeness sake this is with a function:
```rust
match op.node {
    BinOpKind::Le => (lhs, rhs, None),
    BinOpKind::Ge => (rhs, lhs, None),
    BinOpKind::And
        if let ExprKind::Binary(op1, lhs1, rhs1) = lhs.kind
            && let ExprKind::Binary(op2, lhs2, rhs2) = rhs.kind
            && let Some((lt1, gt1)) = normalize_le_ge(op1.node, lhs1, rhs1) 
            && let Some((lt2, gt2)) = normalize_le_ge(op2.node, lhs2, rhs2) =>
    {
        (lt1, gt1, Some((lt2, gt2)))
    },
    _ => return,
}
```

---

With that wall of text out of the way, I do agree that factoring out that bit helps. The reasoning given, however, I disagree with. Sorry for this being as long as it is, I just don't think anything much less than this would give much of a reason for why I don't agree with the reasoning.